heudiconv/convert.py

```diff
@@ -233,6 +233,125 @@ def prep_conversion(sid, dicoms, outdir, heuristic, converter, anon_sid,
+                this_prefix_basename = this_prefix_basename.replace(
+                    label, "_rec-%s%s" % (mag_or_phase, label)
+                )
+                break
+
+    return this_prefix_basename
+
+
+def update_multiecho_name(fileinfo, this_prefix_basename, echo_times):
+    """
+    Insert `_echo-<num>` entity into filename if data are from a multi-echo
+    sequence.
+    """
+    # Field maps separate echoes differently
+    unsupported_types = [
+        '_magnitude', '_magnitude1', '_magnitude2',
+        '_phasediff', '_phase1', '_phase2', '_fieldmap'
+    ]
+    if any(ut in this_prefix_basename for ut in unsupported_types):
+        return this_prefix_basename
+
+    # Get the EchoNumber from json file info.  If not present, use EchoTime
+    if 'EchoNumber' in fileinfo.keys():
+        echo_number = fileinfo['EchoNumber']
+    else:
+        echo_number = echo_times.index(fileinfo['EchoTime']) + 1
+
+    # Determine scan suffix
+    filetype = '_' + this_prefix_basename.split('_')[-1]
+
+    # Insert it **before** the following string(s), whichever appears first.
+    for label in ['_recording', '_proc', '_space', filetype]:
+        if label == filetype:
+            this_prefix_basename = this_prefix_basename.replace(
+                filetype, "_echo-%s%s" % (echo_number, filetype)
+            )
+            break
+        elif (label in this_prefix_basename):
+            this_prefix_basename = this_prefix_basename.replace(
+                label, "_echo-%s%s" % (echo_number, label)
+            )
+            break
+
+    return this_prefix_basename
+
+
+def update_uncombined_name(fileinfo, this_prefix_basename, channel_names):
+    """
+    Insert `_channel-<num>` entity into filename if data are from a sequence
+    with "save uncombined".
+    """
+    # In case any scan types separate channels differently
+    unsupported_types = []
+    if any(ut in this_prefix_basename for ut in unsupported_types):
+        return this_prefix_basename
+
+    # Determine the channel number
+    channel_number = ''.join([c for c in fileinfo['CoilString'] if c.isdigit()])
+    if not channel_number:
+        channel_number = channel_names.index(fileinfo['CoilString']) + 1
+
+    # Determine scan suffix
+    filetype = '_' + this_prefix_basename.split('_')[-1]
+
+    # Insert it **before** the following string(s), whichever appears first.
+    # Choosing to put channel near the end since it's not in the specification yet.
+    for label in ['_recording', '_proc', '_space', filetype]:
+        if label == filetype:
+            this_prefix_basename = this_prefix_basename.replace(
+                filetype, "_channel-%s%s" % (channel_number, filetype)
+            )
+            break
+        elif (label in this_prefix_basename):
+            this_prefix_basename = this_prefix_basename.replace(
+                label, "_channel-%s%s" % (channel_number, label)
+            )
+            break
```

Comment:
duplicate lines got me thinking -- wouldn't it be the same as 
```python
    if (label == filetype) or (label in this_prefix_basename):
            this_prefix_basename = this_prefix_basename.replace(
                label, "_channel-%s%s" % (channel_number, label)
            )
            break
```

Reply:
and it feels like all 3 functions are largely duplicates of the same logic... ok -- I could look into refactoring it all later.  As it came up in https://github.com/nipy/heudiconv/pull/441 we seems to not even have a convenience construct and/or helpers to decompose / recompose (in the standard order defined in BIDS entities table) a standard bids filename into `key: value` pairs.  I feel that we need a basic class such as `BIDSFile` (possibly at large implemented by pybids probably but that one is too heavy ATM to depend on IMHO), which would have a simple dict interface for all the keys, and then additional attribute (e.g. here you call it `.filetype`), and then `__str__` would just return properly formatted filename.  Then checks would become much easier etc.  but ok -- we could do that later I guess

Reply:
Please add a dedicated unit-test for at least one of these functions.

Reply:
Done!